Fix the Installation and the Usage of the LegacyBridge

Plopix opened the pull request:
Not sure about the quality of this fix, but it was not easy to find ;-)

Apparently the _buildLegacyKernelHandlerWeb_ function is called by the Symfony command (through a Listener) and then if we are in CLI we got: 2 errors:

*First*:
```
> Sensio\Bundle\DistributionBundle\Composer\ScriptHandler::clearCache
[Symfony\Component\DependencyInjection\Exception\InactiveScopeException]
  You cannot create a service ("request") of an inactive scope ("request”).
```

Pretty logical: https://github.com/ezsystems/LegacyBridge/blob/master/mvc/Kernel/Loader.php#L157
> The request doesn't exist in CLI.

*Second*:
```
$ php ezpublish/console assets:install --symlink --relative -vvv
  [InvalidArgumentException]
  The target directory "web" does not exist.
```

> Same case here, it's a path issue.

The code below is needed for this 2 issues
```php
if ( PHP_SAPI == 'cli' )
{
     return $this->buildLegacyKernelHandlerCLI();
}
```
But we needed to go back in the good directory instead of the Web Directory (CLI)

ping @andrerom @lolautruche 

++

andrerom commented:
@lolautruche should have a look here, afaik this wasn't the case in the past so probably something caused by the recent refactoring done for sf 2.7.

lolautruche commented:
Mmm that's weird. We fixed this issue a couple of weeks ago. 
What version are you using?

Plopix commented:
That was not that weird after I dived into yesterday.
I'm on the master branch of ezplatform and this one and on Symfony 2.7.2.

But I'm pretty sure that the issue is logical, here is my explanation/story.
Hope you'll find that reading interesting.

When I ran:
```
php ezpublish/console assets:install --symlink --relative -vvv
```
I got this error

```
  [Symfony\Component\DependencyInjection\Exception\InactiveScopeException]
  You cannot create a service ("request") of an inactive scope ("request").
```

```
Exception trace:
 () at /var/www/r2d2html/NOVACTIVE/wwwroot/ezpublish/cache/dev/ezpublishDevDebugProjectContainer.php:17938
 ezpublishDevDebugProjectContainer->getRequestService() at /var/www/r2d2html/NOVACTIVE/wwwroot/vendor/symfony/symfony/src/Symfony/Component/DependencyInjection/Container.php:327
 Symfony\Component\DependencyInjection\Container->get() at /var/www/r2d2html/NOVACTIVE/wwwroot/vendor/ezsystems/legacy-bridge/mvc/Kernel/Loader.php:162
 eZ\Publish\Core\MVC\Legacy\Kernel\Loader->eZ\Publish\Core\MVC\Legacy\Kernel\{closure}() at /var/www/r2d2html/NOVACTIVE/wwwroot/vendor/ezsystems/legacy-bridge/mvc/Kernel/Loader.php:117
 eZ\Publish\Core\MVC\Legacy\Kernel\Loader->eZ\Publish\Core\MVC\Legacy\Kernel\{closure}() at /var/www/r2d2html/NOVACTIVE/wwwroot/vendor/ezsystems/legacy-bridge/mvc/Kernel/Loader.php:338
 eZ\Publish\Core\MVC\Legacy\Kernel\Loader->resetKernel() at /var/www/r2d2html/NOVACTIVE/wwwroot/vendor/ezsystems/legacy-bridge/bundle/EventListener/ConfigScopeListener.php:39
 eZ\Bundle\EzPublishLegacyBundle\EventListener\ConfigScopeListener->onConfigScopeChange() at n/a:n/a
 call_user_func() at /var/www/r2d2html/NOVACTIVE/wwwroot/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/Debug/WrappedListener.php:61
 Symfony\Component\EventDispatcher\Debug\WrappedListener->__invoke() at n/a:n/a
 call_user_func() at /var/www/r2d2html/NOVACTIVE/wwwroot/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php:164
 Symfony\Component\EventDispatcher\EventDispatcher->doDispatch() at /var/www/r2d2html/NOVACTIVE/wwwroot/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/EventDispatcher.php:53
 Symfony\Component\EventDispatcher\EventDispatcher->dispatch() at /var/www/r2d2html/NOVACTIVE/wwwroot/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/ContainerAwareEventDispatcher.php:167
 Symfony\Component\EventDispatcher\ContainerAwareEventDispatcher->dispatch() at /var/www/r2d2html/NOVACTIVE/wwwroot/vendor/symfony/symfony/src/Symfony/Component/EventDispatcher/Debug/TraceableEventDispatcher.php:124
 Symfony\Component\EventDispatcher\Debug\TraceableEventDispatcher->dispatch() at /var/www/r2d2html/NOVACTIVE/wwwroot/vendor/ezsystems/ezpublish-kernel/eZ/Bundle/EzPublishCoreBundle/Console/Application.php:58
 eZ\Bundle\EzPublishCoreBundle\Console\Application->registerCommands() at /var/www/r2d2html/NOVACTIVE/wwwroot/vendor/symfony/symfony/src/Symfony/Bundle/FrameworkBundle/Console/Application.php:73
 Symfony\Bundle\FrameworkBundle\Console\Application->doRun() at /var/www/r2d2html/NOVACTIVE/wwwroot/vendor/ezsystems/ezpublish-kernel/eZ/Bundle/EzPublishCoreBundle/Console/Application.php:46
 eZ\Bundle\EzPublishCoreBundle\Console\Application->doRun() at /var/www/r2d2html/NOVACTIVE/wwwroot/vendor/symfony/symfony/src/Symfony/Component/Console/Application.php:126
 Symfony\Component\Console\Application->run() at /var/www/r2d2html/NOVACTIVE/wwwroot/ezpublish/console:27
```

Then we are in the _buildLegacyKernelHandlerWeb_, but we should not because we are in CLI.
Why?

For me the kernel Closure is called in the Listener: *ConfigScopeListener* (of the LegacyBundle)
This one calls
```php
    public function onConfigScopeChange( ScopeChangeEvent $event )
    {
        $this->kernelLoader->resetKernel();
    }
```
And the function

```php
public function resetKernel()
    {
        /** @var \Closure $kernelClosure */
        $kernelClosure = $this->container->get( 'ezpublish_legacy.kernel' );
        $this->eventDispatcher->dispatch(
            LegacyEvents::PRE_RESET_LEGACY_KERNEL,
            new PreResetLegacyKernelEvent( $kernelClosure() )
        );

        LegacyKernel::resetInstance();
        $this->webHandler = null;
        $this->cliHandler = null;
        $this->restHandler = null;

        $this->container->set( 'ezpublish_legacy.kernel', null );
        $this->container->set( 'ezpublish_legacy.kernel.lazy', null );
        $this->container->set( 'ezpublish_legacy.kernel_handler.web', null );
        $this->container->set( 'ezpublish_legacy.kernel_handler.cli', null );
        $this->container->set( 'ezpublish_legacy.kernel_handler.rest', null );
    }
```


But if I am not wrong: the service: *ezpublish_legacy.kernel* is *ALWAYS* the *WebHandler*.
Because, the handler sent to the factory of buildLegacyKernel is always: ezpublish_legacy.kernel_handler

But:

```yaml
ezpublish_legacy.kernel_handler:
    alias: ezpublish_legacy.kernel_handler.web
```

Then it's why I added the test in PHP CLI in the _buildLegacyKernelHandlerWeb_

> There is maybe another way by fixing the _resetKernel_... but I thought it was better to load the good handler to avoid other side effects.

But with just the 
```php
        if ( PHP_SAPI == 'cli' )
        {
            return $this->buildLegacyKernelHandlerCLI();
        }
```

That was not enough, because for me, there is still another issue.
 Why does the _buildLegacyKernelHandlerCLI_ would place the  path in the _WebDirectory_ ?

I got this error:
```
  [InvalidArgumentException]
  The target directory "web" does not exist.
```

And for sure, when I debuged, the path is already inside /web.. because of the _buildLegacyKernelHandlerCLI_ changed.
The result for instance is that the path for the assets install is not good anymore.

Then I changed:
```php
chdir( $webrootDir ."/../" );
```
 in the _buildLegacyKernelHandlerCLI_

That's it! You know everything ;-)

Let me know!

emodric commented:
Started happening to me again, after upgrading `ezsystems/ezpublish-kernel` from `v6.0.0-alpha3` to `v6.0.0-alpha4`.

emodric commented:
`git bisect` revealed that this PR https://github.com/ezsystems/ezpublish-kernel/pull/1340 is the cause.

andrerom commented:
Yes, I think you are right.
I also think the propoer solution is in progress here :https://github.com/ezsystems/ezpublish-kernel/pull/1366

So with all the french guys on hollidays, feedback there, and some of you testing it, would be highly welcome :)

emodric commented:
@andrerom Didn't help. I did a `git checkout v6.0.0-alpha4` and then `git merge wizhippo/EZP-2469`, but the error is still there with the same exception stack trace.

wizhippo commented:
@emodric @andrerom I just confirmed there is still an issue when you clear the cache.  If you were to run just the console it does work correctly. While clearing the cache the command events are fired in the correct order yet later in the LegacyCachePurger::clear the wrong kernel handler is used.

wizhippo commented:
Looks like the RouterApacheDumperCommand when registering is firing up legacy kernel berfore the commands events get a change to fire and inject the correct cli version.

emodric commented:
Similar to installer command issue I discovered a while back.

wizhippo commented:
During the cache being cleared the changes that LegacyKernelListener applys are undone, so when LegacyCachePurger::clear is called it gets the wrong legacy kernel. @emodric is this this issue you had?  How did you handle it?

wizhippo commented:
Finally found it.  In legacy-bridge ConfigScopeListener resets the the kernel which removes the changes done by the LegacyKernelListener. Not sure how to proceed. I dont whant to repeat the action LegacyKernelListener does in the ConsoleCommandListener

wizhippo commented:
Perhaps LegacyKernelListener needs to also listen to the scope change and apply the same actions as onConsoleCommand or onConsoleCommand could add a scope change event that applys the changes?  I'll stop for now and wait for input.

wizhippo commented:
@emodric for now you can https://gist.github.com/wizhippo/6da884aeeb98fe6b516f

emodric commented:
@wizhippo Thanks, will try it first thing Monday!

yannickroger commented:
ping @emodric Any news on that ?

emodric commented:
@yannickroger Sorry for not reporting it before. I tested with snippet @wizhippo provided and it didin't work. But maybe I misunderstood what to patch exactly...

andrerom commented:
@emodric maybe you can capture @lolautruche, which has been in the wild for three weeks but is expected to show up at summer camp tonight / tomorrow, to discuss this problem and then sync with @wizhippo here on possible solutions?

emodric commented:
@emodric @lolautruche @wizhippo Good idea, will do!

wizhippo commented:
@emodric @lolautruche @wizhippo @Plopix There is the other half of this patch that goes with the legacy bridge.  I just posted it,

https://github.com/ezsystems/LegacyBridge/pull/31

These together are working for me.  Have some concern as to have negative priorities on the listeners.

Hope we can find the best solution soon.

emodric commented:
Thanks @wizhippo I'll try to find some time today to test it. @lolautruche @andrerom @Plopix

lolautruche commented:
Closed in favor of #31
